lightning/src/routing/scoring.rs

```diff
@@ -1549,17 +1579,132 @@ mod approx {
 mod bucketed_history {
 	use super::*;
 
+	// Because liquidity is often skewed heavily in one direction, we store historical state
+	// distribution in buckets of different size. For backwards compatibility, buckets of size 1/8th
+	// must fit evenly into the buckets here.
+	//
+	// The smallest bucket is 2^-14th of the channel, for each of our 32 buckets here we define the
+	// width of the bucket in 2^14'ths of the channel. This increases exponentially until we reach
+	// a full 16th of the channel's capacity, which is reapeated a few times for backwards
+	// compatibility. The four middle buckets represent full octiles of the channel's capacity.
+	//
+	// For a 1 BTC channel, this let's us differentiate between failures in the bottom 6k sats, or
+	// between the 12,000th sat and 24,000th sat, while only needing to store and operate on 32
+	// buckets in total.
+
+	const BUCKET_START_POS: [u16; 33] = [
+		0, 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 3072, 4096, 6144, 8192, 10240, 12288,
+		13312, 14336, 15360, 15872, 16128, 16256, 16320, 16352, 16368, 16376, 16380, 16382, 16383, 16384,
+	];
+
+	// To figure out which bucket a given 2^14th is in quickly, we use a lookup table. However,
+	// because doing a full lookup table would require 16KiB, we do a split lookup table, with
+	// values above the 64th position in _B where the pos is divided by 2^7.
+	const POS_TO_BUCKET_A: [u8; 64] = [
+		 0,  1,  2,  2,  3,  3,  3,  3,  4,  4,  4,  4,  4,  4,  4,  4,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,  5,
+		 6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,  6,
+	];
+	const POS_TO_BUCKET_B: [u8; 64] = [
+		 7,  8,  9,  9, 10, 10, 10, 10, 11, 11, 11, 11, 11, 11, 11, 11, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 13, 13, 13, 13, 13,
+		14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+	];
+
+	const LEGACY_TO_BUCKET_RANGE: [(u8, u8); 8] = [
+		(0, 12), (12, 14), (14, 15), (15, 16), (16, 17), (17, 18), (18, 20), (20, 32)
+	];
+
+	const MIN_SIZE_BUCKETS: u16 = 1 << 14;
```

Comment:
nit: I feel like this might be better renamed to something like `TOTAL_BUCKET_UNITS`, since it's mostly used as the max "position" or the total number of smallest buckets

Reply:
SGTM